### PR TITLE
Sunset RLS

### PR DESCRIPTION
**Pull request by ehuss**
This removes RLS per the plan outlined in https://blog.rust-lang.org/2022/07/01/RLS-deprecation.html. This replaces the `rls` executable with a small program which will display an alert telling the user that RLS is no longer available.

An overview of the changes here:
* Removes the rls submodule and replaces it with a small stub program.
* `rls` is removed from `./x.py install`. I do not think users running `install` will need the stub.
* `rls` is removed from `./x.py test`, it doesn't have any tests.

Other things of note:
* I kept `DIST_REQUIRE_ALL_TOOLS` even though it is no longer needed, with the thought that it could be useful in the feature. However, I could remove it if desired.
* I kept `extra_deps` in `tool_extended` (which allows tools to depend on other things), even though it is no longer needed. This can also be removed if desired.
* ~~This keeps RLS in the macOS `pkg` installer and the Windows `msi` installer. I kinda lean towards removing it from those, but I'm not sure?~~ RLS has been removed from pkg and msi.
* This does not remove the analysis component. It is not clear if we should keep this or not. RLS was the primary user, but I think there are a few users that have made tools around it.
* There will be several followup steps after this PR:
    * Updating the toolstate repo to remove RLS.
    * Updating documentation, such as rustc-dev-guide, to remove references to RLS.

The alert looks like this in VSCode:

image

In Sublime it looks similar.

I would appreciate if others could help test with other editors such as vim or Emacs.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -4835,18 +4467,18 @@ dependencies = [
...
 [[package]]
 name = "serde"
-version = "1.0.140"
+version = "1.0.143"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "fc855a42c7967b7c369eb5860f7164ef1f6f81c20c7cc1141f2a604e18723b03"
+checksum = "53e8e5d5b70924f74ff5c6d64d9a5acd91422117c60f48c4e07855238a254553"
 dependencies = [
  "serde_derive",
 ]
```
Comment (klensy) [inline Cargo.lock:4475]: Few updates here and lower sneaked in, but i guess it's ok?

File: src/tools/rustc-workspace-hack/Cargo.toml [hunk 2]
```diff
@@ -72,25 +73,16 @@ features = [
...
 # Ensure default features of regex, which are disabled in some scenarios.
 regex = { version = "1.5.6" }
-proc-macro2 = { version = "1", features = ["default"] }
-quote = { version = "1", features = ["default"] }
-rand_core_0_5 = { package = "rand_core", version = "0.5.1", features = ["getrandom", "alloc", "std"] }
-serde = { version = "1.0.82", features = ['derive'] }
 serde_json = { version = "1.0.31", features = ["raw_value", "unbounded_depth"] }
-smallvec = { version = "1.8.1", features = ['union', 'may_dangle'] }
-syn = { version = "1", features = ['fold', 'full', 'extra-traits', 'visit', 'visit-mut'] }
 url = { version = "2.0", features = ['serde'] }
```
Comment (klensy) [inline src/tools/rustc-workspace-hack/Cargo.toml:86]: Probably not all removed crated should be removed, as some of them still used somewhere else.

**Comment by klensy** [comment]
Also this can be removed?
https://github.com/rust-lang/rust/blob/1cff5642037b83ce1239a624bbe617a9aa0d59b1/compiler/rustc_driver/src/lib.rs#L163-L180

File: src/tools/rls/LICENSE-MIT [hunk 3]
```diff
@@ -0,0 +1,23 @@
+Permission is hereby granted, free of charge, to any
```
Comment (Mark-Simulacrum) [inline src/tools/rls/LICENSE-MIT:1]: Hm, are these files needed? Seems like the code would be licensed under the root LICENSE-* files anyway, which should be ~equivalent?

File: src/tools/rustfmt/atom.md [hunk 4]
```diff
@@ -1,8 +1,8 @@
 # Running Rustfmt from Atom
 
-## RLS
+## rust-analyzer
 
-Rustfmt is included with the Rust Language Server, itself provided by [ide-rust](https://atom.io/packages/ide-rust).
+Rustfmt is included with rust-analyzer, itself provided by [ide-rust](https://atom.io/packages/ide-rust).
```
Comment (Mark-Simulacrum) [inline src/tools/rustfmt/atom.md:5]: Is this true today? I think rust-analyzer shells out to rustfmt, doesn't directly inline it (otherwise can't build on stable)?

I would expect users to need to rustup component add rustfmt or so.

**Review by Mark-Simulacrum — COMMENTED** [review]
This looks largely good to me; I think leaving further cleanups of the more interesting decisions (e.g., DIST_REQUIRE_ALL_TOOLS, other things listed in your list) to future PRs should be fine. Definitely great to see a bunch of simplifications.

I think dropping RLS from the pkg and msi installers in this PR makes sense though; usage on those is probably quite low and I'm not sure an "upgrade" there is typically done without checking back to blog posts etc (unlike may be the case with rustup).

A few smaller comments to resolve left inline.

File: src/tools/rustfmt/atom.md (hunk 4 above)
Comment (calebcartwright) [reply src/tools/rustfmt/atom.md:5]: Good point, and yes that's correct as per my understanding as well. RLS was a rustfmt-as-a-lib consumer, but RA invokes rustfmt directly using the stdin mode to pipe in file/snippet contents.

I'm not sure how much utility these docs provide TBH as they've not really been kept up to date, but if we want to make an update perhaps something to the effect of:

```markdown
Rustfmt can be utilized from rust-analyzer,...
```

File: src/tools/rls/LICENSE-MIT (hunk 3 above)
Comment (ehuss) [reply src/tools/rls/LICENSE-MIT]: Sure! I have changed it to use the root files instead.

File: src/tools/rustfmt/atom.md (hunk 4 above)
Comment (ehuss) [reply src/tools/rustfmt/atom.md]: Good catch!  I was not reading it closely. I have tweaked the wording so it does not imply that rust-analyzer installs rustfmt.

File: src/tools/rustc-workspace-hack/Cargo.toml (hunk 2 above)
Comment (ehuss) [reply src/tools/rustc-workspace-hack/Cargo.toml:86]: These were removed based on extensive testing to ensure they aren't used elsewhere.  There is some possibility that one of the more obscure platforms may enable one of these features, but that can be resolved via CI.

Comment (klensy) [reply src/tools/rustc-workspace-hack/Cargo.toml:86]: Uhh, i don't understand. For example, `proc-macro2` with friends of `quote` and `syn` used literally everywhere: rustc\rustdoc\clippy\cargo\rustfmt.

Yes, removing this entries probably don't break things now, until some of crates decide to add\remove some of features (at which point CI will fail and contributor will be forced to return back here and rollback something).

Comment (ehuss) [reply src/tools/rustc-workspace-hack/Cargo.toml:86]: That is true of any shared dependency between the tools. Features can be adjusted at any time. The rustc-workspace-hack doesn't cover all of the shared dependencies (nor should it, since not all tools use the same dependencies).  I don't think rustc-workspace-hack should proactively add dependencies in the anticipation that they might be needed in the future. Adding new entries is relatively easy.  

As for proc-macro2 and quote, I doubt those will ever appear without their default features. It is not really clear from the history why those were added.

**Comment by ehuss** [comment]
> think dropping RLS from the pkg and msi installers in this PR makes sense though;

Pushed an update that removes it from the msi and pkg installers.

**Comment by bjorn3** [comment]
> Also this can be removed?
> 
> https://github.com/rust-lang/rust/blob/1cff5642037b83ce1239a624bbe617a9aa0d59b1/compiler/rustc_driver/src/lib.rs#L163-L180

No, I don't think so. While RLS is the main user, it is also useful outside of RLS. For example my (not updated for current nightly) https://github.com/bjorn3/rust_read_rlib uses `set_file_loader` to generate a fake `lib.rs` without having to write anything to the disk. They may also be used to fix https://github.com/rust-lang/rust-analyzer/issues/3107 at some point in the future.

**Comment by ehuss** [comment]
I fixed the `syn` issue I believe. There is something fishy with how the duplicate checks are working with cross-compiling (build-dependencies don't seem to be checked otherwise). I may investigate that more later.

@bors r=Mark-Simulacrum

**Comment by eddyb** [comment]
> This does not remove the analysis component. It is not clear if we should keep this or not. RLS was the primary user, but I think there are a few users that have made tools around it.

First off: is there a tracking issue that we could discuss this on? I will move this comment wherever necessary.

<hr/>

I believe [searchfox's Rust indexer](https://github.com/mozsearch/mozsearch/blob/e5de767b4d47bd40cf12596943c15220bd09100f/tools/src/bin/rust-indexer.rs) might be one of the more important users but I don't really know much about it, sadly (@emilio's handle shows up in the comments, maybe they can shed some light?).

My rough suggestion would be to look at what information it reads and try to reduce the `-C save-analysis` implementation in the compiler so that it doesn't outright break, but we put less effort into "offering detail".
(imagine things like pretending `module::function` and `Type::method` are the same thing, or that there is no difference between `enum`/`struct`s and `type` aliases, etc.)

I would expect we may be able to shrink it all the way to "dump all use->def relations where we have unique `Span` mappings" style approach (roughly structured like a lint over the HIR that wants to see both direct paths, and inferred e.g. method callees, using the "current typeck results" pattern available to lints, though other passes also mimic their own - `rustc_privacy` would be a great example if it were much smaller).

AFAIK, this information would also be very similar to what `rustdoc` source annotation needs (cc @GuillaumeGomez @Manishearth), so there's maybe some sharing opportunity there? (e.g. make a query computing something like this, `-C save-analysis`' only job becomes dumping it into some JSON format)

**Comment by bjorn3** [comment]
Rust-analyzer has support for emitting LSIF formatted analysis results as well as Sourcegraph's SCIP. This might be a viable alternative for the save-analysis results for most users.

**Comment by ehuss** [comment]
I have no intent on making changes to save-analysis. If you want to start a discussion on that, I would imagine #43606 or something like an MCP or internals discussion would be a good start.

**Comment by eddyb** [comment]
Ah, thanks for the keywords @bjorn3, I was able to find LSIF mentioned twice already wrt `searchfox`:
* https://github.com/mozsearch/mozsearch/pull/81#issuecomment-1079986893
* https://github.com/mozsearch/mozsearch/pull/314#issuecomment-1008431841

Based on https://bugzilla.mozilla.org/show_bug.cgi?id=1761287 I'm guessing we should coordinate with @asutherland and eventually remove `save-analysis` in favor of RA+LSIF.

Will continue this discussion #43606 (saw @ehuss' comment as I was writing this, heh).

**Comment by Xanewok** [comment]
Thanks a lot @ehuss, appreciate it!
